### ISSUE TITLE
Use External Event instead of idle

### ISSUE THREAD

**Issue by Frederik91** [mcneel/rhino.inside-revit#319]
Hi

I noticed that you are subscribing to the Idle-event to get Revit context. As described in this blog this is not recommended, reasoned as following: 

> f you want to use it for continuous polling, it gives you absolutely no control over the frequency. Revit will call it as often as it can, hogging the entire CPU time and bringing performance down.

[https://thebuildingcoder.typepad.com/blog/2015/12/external-event-and-10-year-forum-anniversary.html](https://thebuildingcoder.typepad.com/blog/2015/12/external-event-and-10-year-forum-anniversary.html)

As a Revit addin developer myself I did the very same thing as you. Fortunatly it's not very complicated to use the ExternalEvent instead of idle-event, so I'll argue it's worth the effort.

**Comment by kike-garbo** [comment]
Hi,

Thanks for your warning.
We are using `Idling` event mainly for previews generation rest of the UI is managed by an `ExternalEvent`.
Some times those previews are expensive to generate and is just not acceptable to make the user wait until the previews are generated to let him use Revit.
Unfortunately those previews can only be generated on the main thread while Revit is refreshing the view. So to make Revit feel interactive while the previews are generated we use the Idling event that is only fired when there is no user input on the queue, this means no key strokes nor mouse events. At that time we start building the previews in slices of 200 ms. when this time is over we return back control to Revit. If at that moment there are user events in queue Revit will process those events, if not it is idle and will fire the Idling event again. When all previews are generated, so no more work is pending, we stop requesting the Idling event and this loop stops and CPU goes again to a waiting state.

Using a totally synchronous implementation the user should wait for the preview, some times more than seconds witch is not acceptable for a preview, with this other implementation he can manipulate the model or even start other commands while the previews are generated. In both cases the amount of work is the same and the amount of time the CPU is working is the same.

If we go a bit deeper on how `ExternalEvent` is implemented we can see that it is implemented on top of `Idling` event.
It is mainly a `Queue` of `ExternalEvent` instances that are enqueued when you call `Rise` on them and fired, in Addin context, during Idling event in queue order, nothing fancier.
The main advantage of `ExternalEvent` over Idling event is that events are dispatched in rising order witch is mandatory for UI handling. This can not be achieved with Idling event that can only be called in register order.

So in summary, my experience is, for UI events `ExternalEvent` is a must to enforce rising order, but for background-idling stuff you still need Idling event.
